src/utils/permissions-utils.js

```diff
@@ -19,3 +19,129 @@ export function canMove(entity) {
+  }
+  const authorizedIndices = authorizedSchemas[template];
+  for (const index in data.components) {
+    if (!data.components.hasOwnProperty(index)) continue;
+    if (!authorizedIndices.includes(index)) {
+      data.components[index] = null;
+    }
+  }
+  return data;
+}
+
+function authorizeEntityManipulation(entity, sender, senderPermissions) {
+  const { template, creator } = entity.components.networked.data;
+  const isPinned = entity.components.pinnable && entity.components.pinnable.data.pinned;
+  const isCreator = sender === creator;
+
+  if (template.endsWith("-avatar")) {
+    return isCreator;
+  } else if (template.endsWith("-media")) {
+    return (!isPinned || senderPermissions.pin_objects) && (isCreator || senderPermissions.spawn_and_move_media);
+  } else if (template.endsWith("-camera")) {
+    return isCreator || senderPermissions.spawn_camera;
+  } else if (template.endsWith("-pen")) {
+    return isCreator || senderPermissions.spawn_pen;
+  } else {
+    return false;
+  }
+}
+
+function authorizeOrSanitizeMessageData(data, sender, senderPermissions) {
+  const entity = NAF.entities.getEntity(data.networkId);
+  if (!entity) return false;
+
+  if (authorizeEntityManipulation(entity, sender, senderPermissions)) {
+    return true;
+  } else {
+    const { template } = entity.components.networked.data;
+    sanitizeMessageData(template, data);
+    return true;
+  }
+}
+
+const emptyObject = {};
+export function authorizeOrSanitizeMessage(message) {
+  const { dataType, from_session_id } = message;
+
+  if (dataType === "u" && message.data.isFirstSync && !message.data.persistent) {
+    // The server has already authorized first sync messages that result in an instantiation.
+    return message;
+  }
+
+  const senderPermissions = from_session_id
+    ? window.APP.hubChannel.presence.state[from_session_id].metas[0].permissions
+    : null;
+
+  if (dataType === "um") {
+    for (const index in message.data.d) {
+      if (!message.data.d.hasOwnProperty(index)) continue;
+      const authorizedOrSanitized = authorizeOrSanitizeMessageData(
+        message.data.d[index],
+        from_session_id,
+        senderPermissions
+      );
+      if (!authorizedOrSanitized) {
+        message.data.d[index] = emptyObject;
+      }
+    }
+    return message;
+  } else if (dataType === "u") {
+    const authorizedOrSanitized = authorizeOrSanitizeMessageData(message.data, from_session_id, senderPermissions);
+    if (authorizedOrSanitized) {
+      return message;
+    } else {
+      return emptyObject;
+    }
+  } else if (dataType === "r") {
+    const entity = NAF.entities.getEntity(message.data.networkId);
+    if (!entity) return emptyObject;
+    if (authorizeEntityManipulation(entity, from_session_id, senderPermissions)) {
+      return message;
+    } else {
+      return emptyObject;
+    }
+  } else {
+    // Pass through for other data types. Namely, "drawing-<networkId>" messages
+    return message;
```

Comment:
seems like in this context we should whitelist + reject unknown message types

Reply:
We decided not to bother with a whitelist on the server side either. I think it's safe not to bother with a whitelist for `dataTypes`, since they need explicit subscribers already.